javascript/implementation/KeySet.ts

```diff
@@ -0,0 +1,228 @@
+import { GinkInstance } from "./GinkInstance";
+import { Container } from "./Container";
+import { KeyType, Muid, AsOf } from "./typedefs"
+import { Bundler } from "./Bundler";
+import { ensure, muidToString, muidTupleToMuid } from "./utils";
+import { interpret, toJson } from "./factories"
+import { Behavior, ContainerBuilder } from "./builders";
+
+export class KeySet extends Container {
+
+    constructor(ginkInstance: GinkInstance, address: Muid, containerBuilder?: ContainerBuilder) {
+        super(ginkInstance, address, Behavior.KEY_SET);
+        if (this.address.timestamp < 0) {
+            ensure(address.offset == Behavior.KEY_SET);
+        } else {
+            ensure(containerBuilder.getBehavior() == Behavior.KEY_SET);
+        }
+    }
+
+    /**
+     * Adds a key to the keyset.
+     * If a bundler is supplied, the function will add the entry to that bundler
+     * and return immediately (presumably you know what to do with a CS if you passed it in).
+     * If the caller does not supply a bundler, then one is created on the fly, and
+     * then this method will await on the CS being added to the database instance.
+     * This is to allow simple console usage like:
+     *      await myKeySet.add("foo");
+     * @param key
+     * @param change an optional bundler to put this in.
+     * @returns a promise that resolves to the address of the newly created entry
+     */
+    async add(key: KeyType, change?: Bundler|string): Promise<Muid> {
+        return await this.addEntry(key, Container.INCLUSION, change);
+    }
+
+    /**
+     * Similar to add method, but for multiple entries.
+     * @param keys an iterable of keys to add to the key set
+     * @param change an optional bundler to put this in.
+     * @returns a promise that resolves to a set of Muids for the created entries.
+     */
+    async update(keys: Iterable<KeyType>, change?: Bundler|string): Promise<Set<Muid>> {
+        let bundler: Bundler;
+        if (change instanceof Bundler) {
+            bundler = change;
+        } else {
+            bundler = new Bundler(change);
+        }
+        const additions = new Set<Muid>;
+        for (const key of keys) {
+            additions.add(await this.addEntry(key, Container.INCLUSION, bundler));
+        }
+        await this.ginkInstance.addBundler(bundler);
+        return additions;
+    }
+
+    /**
+     * Tries to delete the key and return the key itself, throws an error if key not found.
+     * @param key one key of KeyType
+     * @param change an optional bundler to put this in.
+     * @returns a promise that resolves to a key.
+     */
+    async pop(key: KeyType, change?: Bundler|string): Promise<KeyType> {
```

Comment:
There's no pop function for the Javascript Set so there shouldn't be one here either.